core/src/main/scala/org/apache/spark/sql/streaming/SnappySinkCallback.scala

```diff
@@ -0,0 +1,184 @@
+
+import io.snappydata.StreamingConstants._
+import org.apache.log4j.Logger
+
+import org.apache.spark.sql.execution.streaming.Sink
+import org.apache.spark.sql.sources.{DataSourceRegister, StreamSinkProvider}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, Dataset, Row, SnappySession, _}
+import org.apache.spark.util.Utils
+
+/**
+ * Should be implemented by clients who wants to override default behavior provided by
+ * [[DefaultSnappySinkCallback]].
+ * <p>
+ * In order to override the default callback behavior the qualified name of the implementing
+ * class needs to be passed against `sinkCallback` option while defining stream query.
+ *
+ */
+trait SnappySinkCallback {
+
+  /**
+   * This method is called for each streaming batch after checking the possibility of batch
+   * duplication which is indicated by `possibleDuplicate` flag.
+   * <p>
+   * A duplicate batch might be picked up for processing in case of failure. In case of batch
+   * duplication, this method should handle batch in idempotent manner in order to avoid
+   * data inconsistency.
+   */
+  def process(snappySession: SnappySession, sinkProps: Map[String, String],
+              batchId: Long, df: Dataset[Row], possibleDuplicate: Boolean = false): Unit
+}
+
+class SnappyStoreSinkProvider extends StreamSinkProvider with DataSourceRegister {
+
+  @Override
+  def createSink(
+                  sqlContext: SQLContext,
+                  parameters: Map[String, String],
+                  partitionColumns: Seq[String],
+                  outputMode: OutputMode): Sink = {
+    createSinkStateTableIfNotExist(sqlContext)
+    val cc = try {
+      Utils.classForName(parameters(SINK_CALLBACK)).newInstance()
+    } catch {
+      case _: NoSuchElementException => new DefaultSnappySinkCallback()
+    }
+
+    SnappyStoreSink(sqlContext.asInstanceOf[SnappyContext].snappySession, parameters,
+      cc.asInstanceOf[SnappySinkCallback])
+  }
+
+  private def createSinkStateTableIfNotExist(sqlContext: SQLContext) = {
+    sqlContext.asInstanceOf[SnappyContext].snappySession.sql(s"create table if not exists" +
+      s" $SINK_STATE_TABLE (" +
+      " stream_query_id varchar(200)," +
+      " batch_id long, " +
+      " PRIMARY KEY (stream_query_id)) using row options(DISKSTORE 'GFXD-DD-DISKSTORE')")
+  }
+
+  @Override
+  def shortName(): String = SNAPPY_SINK_NAME
+
+}
+
+case class SnappyStoreSink(snappySession: SnappySession,
+    parameters: Map[String, String], sinkCallback: SnappySinkCallback) extends Sink {
+
+  override def addBatch(batchId: Long, data: Dataset[Row]): Unit = {
+    val streamQueryId = parameters(STREAM_QUERY_ID).toUpperCase
+
+    val updated = snappySession.sql(s"update $SINK_STATE_TABLE " +
+      s"set batch_id=$batchId where stream_query_id='$streamQueryId' and batch_id != $batchId")
+      .collect()(0).getAs("count").asInstanceOf[Long]
+
+    // TODO: use JDBC connection here
+    var posDup = false
```

Comment:
Come to think of it, using JDBC connection is a required here. We have to use it with transaction isolation level as READ_COMMITTED or better REPEATABLE_READ. There is no way to set those from Spark APIs. Without those there are no atomicity guarantees on any changes on the row table. Use a pooled connection like in JDBCMutableRelation.executeUpdate (but set and reset the transaction isolation level on the connection explicitly).

Reply:
Is atomicity required here? There is only one thread modifying one row in the table and even the reader thread is same. Unless we want to take care of inconsistency due to all node failures, simple operation should suffice.

Anyways, we have plans to use connection later for performance reasons as suggested earlier. Will track that as JIRA ticket.

Reply:
There are some corner cases where replicated table apply can result in inconsistencies. For instance if lead node fails after sending the update to few nodes but before sending to remaining, then replicas can become inconsistent. Even eventual consistency will not help. Most of these cases are handled for partitioned regions even without transactions, but for replicated tables we do need transactions for atomicity and consistency. READ_COMMITTED is enough for single row changes.

Reply:
Added a JIRA to track this: SNAP-2598